### ISSUE TITLE
Choosing the right YOLO

### ISSUE THREAD

**Issue by linab8** [FabianPlum/OmniTrax#44]
Hello @FabianPlum,
I have two questions. First, before using Omnitrax on our original data, I'm trying to replicate your tracking model on "example_ant_recording". I'm using "yolov4-big_and_small_ants_480.cfg", as well as "yolov4-big_and_small_ants_1024px_refined_with_2048_px_27000.weights", but I wanted to check what values were used for other parameters in the tracking settings with which you got this video output: https://github.com/FabianPlum/OmniTrax/blob/main/images/example_ant_tracked.gif ?
I want to replicate this first so I can better understand how modifying tracking parameters affects the outcome before continuing to our original data.

Regarding our data, I'm unsure how to determine which YOLO is the best fit. As far as I understand, your pre-trained networks are in the atta_single_class folder. For example, our videos are the most similar to your "desert termites" video regarding the soil background, but we have polymorphic ants, so I'm trying to understand how to decide which of your pre-trained networks we could use.

Best regards,
Lina

**Comment by FabianPlum** [comment]
Hi @linab8 

Could you perhaps share some example video footage with me and could you let me know which species you are working with?

All the best
Fabi

**Comment by linab8** [comment]
@FabianPlum 
Here's a cropped video as an example. The species is _Messor wasmanni_ and we have a variable background between videos - we'll be using multiple videos of multiple colonies from different locations (ranging from naked soil background to more vegetation, etc.)

https://drive.google.com/file/d/1YifpjQWLy1S_LjPIZ1bBTbgfCB7xc49F/view?usp=sharing

Best,
Lina

**Comment by FabianPlum** [comment]
Hi Lina,

I just had a look at the video and I'm not sure OmniTrax is the perfect tool for you here, at least not without refinement training on your specific data appearance, or getting footage much closer to the individuals. The YOLO model(s) we use with OmniTrax require the ants to identifiable as such in each frame, rather than through movement in continuous frames. For a human annotator looking at the flow of individuals it may seem trivial to identify what is part of the background and what is part of an ant but in this case, for a neural network to differentiate cleanly between ant and noise without movement info is quite hard. 

If you cannot easily produce footage at higher resolutions or, ideally simply closer to the individuals, using a different tracking approach may be the better option. Perhaps have a look at [https://trex.run/](https://trex.run/) or [https://idtracker.ai/latest/](https://idtracker.ai/latest/).

They rely more strongly on motion and background subtraction to account for textured but static backgrounds.

If you you want to proceed with OmniTrax and don't have a way to easily get footage under different recording conditions, we can look into training custom models for you. That, however, will also come with some hand annotation required for each of your recording scenarios, so we can train a model specifically for those purposes. 

All the best
Fabi

**Comment by markobracic** [comment]
Hi Fabi,
I am working with Lina on training to implement OmniTrax for following ants in the field. First, thanks for developing and for offering support!

Considering the footage, you recommend getting closer, with the expense of distance tracked? We were concerned that there would be too few frames where an individual is in the video to calculate, e.g. speed properly. And we use GoPRo 11 with a minimum focal distance of 30cm so ants become slightly blurry if we move closer. Here is an example from ~20cm:
https://www.dropbox.com/scl/fi/9tw9tw0ix0zby9nhwwa9k/GX010016.MP4?rlkey=5yqjc57w5sv5jh90whe1m1pb6&dl=0

Could that help? Or we need a cam that can focus closer?

I thought that trex.run and idtracker are not doing great when there is no contrasting background but we try, thx!

**Comment by FabianPlum** [comment]
Hi @markobracic!
YES! The footage is going very much in the right direction. If we got the same footage, perhaps slightly closer still, but with the indidivuals in focus, we could probably get very decent results.

I went ahead and tested OmniTrax with the [default single class YOLOv4 model trained on synthetic data](https://drive.google.com/drive/folders/1PSseMeClcYIe9dcYG-JaOD2CzYceiWdl) and increased the network size in the detection configuration. That makes tracking slower but ensures individuals that are small relative to the image size can be picked up.

![Image](https://github.com/user-attachments/assets/d94f7814-f161-4e08-8d4c-3b7e34fc039a)

It's far from perfect, but certainly a step in the right direction:

https://github.com/user-attachments/assets/04d59a01-73cf-4bdc-9659-c915408cd691

I used the [masking option](https://github.com/FabianPlum/OmniTrax/blob/main/docs/tutorial-tracking.md#masking-regions-of-interest-optional) to mask out the bottom of the frame, because I had to decrease the confidence threshold quite significatly which leads to more false positives to be picked up.

I usually film outdoors footage with a DSLR (or mirrorless camera) where I can precisely control for lighting settings as well, especially with high shutter-speeds being crucial to cut out motion blur.

**Comment by markobracic** [comment]
Thanks a lot for the test! It looks like there is hope for us after all! 🥳 

We are out for the fieldwork next week and we will try to get footage from different heights. It would be great if you could help us pick the most promising setup. 

If this approach fails, we are planning to manually label some individuals to obtain some output so we can use that to train the model for the future. We will be more than happy to collaborate with you!

Thanks for the camera tip! I got the info that GoPro should be good for tracking.

**Comment by FabianPlum** [comment]
Hi @markobracic and @linab8!
How are things going on the recording front?

All the best
Fabi

**Comment by linab8** [comment]
Hi @FabianPlum ! We're currently doing field work, but we tried out different heights with GoPro last week (there weren't a lot of ants active yet). Unfortunately, when we get closer to the ground with GoPro, ants are out of focus, as GoPro12's minimal focus length is 30 cm. Here is an example of a video filmed from 15 cm height, in 4K. We're looking into macro lenses for the GoPro, or even better, a digital camera with an incorporated macro mode. Would this height of 15 cm be alright if we could get the ants into focus with another camera? Also, as you can see, we have a lot of vegetation in this video, so it might be more problematic than the previous video.

https://drive.google.com/file/d/1EXzA_mWYkUpY8MCEz9MH2Kze7x8KvIGz/view?usp=sharing

**Comment by linab8** [comment]
Hello @FabianPlum ! We're currently between field trips, so we're looking at potential equipment for the future. Could you advise us on some camera options? I'm considering an Olympus camera, but I'm not sure whether its specifications are good enough for filming behaviour in the field. This is the model: https://www.amazon.com/OM-SYSTEM-Tough-TG-7-Digital/dp/B0CH1QPT77?th=1
What would you recommend?

**Comment by FabianPlum** [comment]
Hi Lina,
How are you doing?
Sorry for the long radio silence, I believe you emailed me somewhere as well, but I can't seem to find the message, so let's pick up things here!

As for the camera you mention above, I have never used it before, so I can't provide a direct recommendation. If you can record videos with it at a reasonable magnification and at a framerate that is adequate for your purposes, it'll probably be fine.

More importantly, you have to think about how you are going to shoot videos. The tripod top-down setting is a good one but there are a few things you can try to get the best possible footage:

- bring external light sources, the more light the better. We need to make sure that we can bring down the exposure time as much as possible as motion blur will kill the accuracy of the tracker. If, in a single frame, you can't make out what is and sin't an ant, the trained yolo model won't be able to do that either
- choose a high contrast background, if possible. If the ants blend in with the background it becomes harder to detect them well, and lowering the confidence threshold too much will lead to false-positives being picked up, i.e. parts of the background will be mistakenly identified as ants and may show up in your tracking data

I think, to make your tracking more robust, it would be a good idea to actually retrain the detection network on some of your footage, once you have collected it. If the goal is to collect footage in the wild and then analyse it later, I can help you refine the neural network on your data specifically so you can make a stronger detector which will lead to much better tracking results. The procedure is relatively simple, it just takes a while to set up, so I would offer to train the neural networks for you if you can annotate the images. Essentially, you will just need to track a few minutes of your collected footage with OmniTrax, export the tracking data and then I can help you re-train the yolo detector to learn how the ants appear in your footage.

Hope this helps!
I will now update the OmniTrax git more frequently again, so please shoot me another message here with any questions you have!

All the best
Fabi
